### PR TITLE
Iot2000 ebg swu

### PR DESCRIPTION
**Pull request by andi8086**
# EFIBootGuard and SWUpdate #

These commits update the existing `meta-iot2000` layer to use integrated `wic` and `kas` build tools.
The boot loader is changed to `EFI Boot Guard` and some fixes are applied. The 2nd half of the commits integrate `SWUpdate` to support system updates.

### CODE REVIEW

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh [hunk 1]
```diff
@@ -1,31 +1,29 @@
 #!/bin/sh
 
-set -eu
```
Comment (stormc) [inline meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:3]: Why did you remove this? At least `-u` has some value.

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh [hunk 2]
```diff
@@ -1,31 +1,29 @@
...
+	update-rc.d -f expandfs.sh remove
+}
 
-START_BLOCK=$(parted $ROOT_DEVICE -ms unit s p | grep "^$PART_NUMBER" | cut -f 2 -d: | sed "s/s$//")
+ROOT_DEVICE=$(findmnt / -o source -n | sed 's/p\?[0-9]*$//')
+LAST_PART=$(sfdisk -d ${ROOT_DEVICE} | tail -1 | awk '{print $1}')
 
-parted -ms $ROOT_DEVICE rm $PART_NUMBER
+log "Resizing last partition"
+sfdisk -d ${ROOT_DEVICE} 2>/dev/null | grep -v last-lba | \
```
Comment (stormc) [inline meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: Are you sure the locale is actually EN or is, e.g., the DE variant also printing "last-lba"?

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh [hunk 3]
```diff
@@ -1,31 +1,29 @@
 #!/bin/sh
 
-set -eu
+THIS_SCRIPT="$(basename $0)"
 
-PART_NUMBER=3
+log () {
+	echo "${THIS_SCRIPT}: $1" > /dev/kmsg
```
Comment (stormc) [inline meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: Below, you use `logger`. Why not also here or just one of both?

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh [hunk 4]
```diff
@@ -1,31 +1,29 @@
...
+LAST_PART=$(sfdisk -d ${ROOT_DEVICE} | tail -1 | awk '{print $1}')
 
-parted -ms $ROOT_DEVICE rm $PART_NUMBER
+log "Resizing last partition"
+sfdisk -d ${ROOT_DEVICE} 2>/dev/null | grep -v last-lba | \
+	sed 's|\('${LAST_PART}' .*, \)size=[^,]*, |\1|' | \
+	sfdisk --force ${ROOT_DEVICE} 2>&1 | logger -t ${THIS_SCRIPT}
 
-parted -ms $ROOT_DEVICE unit s -- mkpart primary ext3 $START_BLOCK -1
+log "Informing kernel about new partitioning"
```
Comment (stormc) [inline meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: Doubly logged disabling, once here, once in the function.

File: meta-iot2000-example/recipes-core/images/iot2000-example-image.inc [hunk 5]
```diff
@@ -43,3 +43,6 @@ IMAGE_INSTALL_append = " ofono ofono-tests"
 IMAGE_INSTALL_append = " connman connman-tools connman-tests connman-client"
 
 DEPENDS += "gptfdisk-native efibootguard-native"
+
+# The following is needed for correct EFI deployment
```
Comment (stormc) [inline meta-iot2000-example/recipes-core/images/iot2000-example-image.inc]: Yes, but the reason in code would be nice as well : )

File: meta-iot2000-example/README.SWUPDATE.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
+# Example for efibootguard + SWUpdate support with meta-iot2000
+
+[SWUpdate](https://github.com/sbabic/swupdate) is a software that supports
+firmware and/or software updates targeting embedded devices.
+
+It can be run as a stand-alone server in foreground or as a daemon listening
+for connections via ethernet. Different backends can be configured to support
```
Comment (stormc) [inline meta-iot2000-example/README.SWUPDATE.md]: ... ethernet or IPC

File: meta-iot2000-example/README.SWUPDATE.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+default it installs an init script to start up a webserver for update purposes,
+which is not needed for manual updates. For this purpose, it is replaced by an
+empty init script.
+
+This example of `SWUpdate` and `efibootguard` integration provides manual
+update capability only.
+
+
+The manual update process means, that the `swu` file must be copied onto the
+finished SD-card. To use it with swupdate, the following command is used:
```
Comment (stormc) [inline meta-iot2000-example/README.SWUPDATE.md]: .. or via SSH or something to the booted device. The setup can be adapted to work with backends like, e.g., hawkBit

File: meta-iot2000-example/recipes-support/swupdate/swupdate/swupdate.cfg [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+{
+    verbose  = true;
+    loglevel = 10;
+    syslog   = false;
+};
+
+processes :
+(
+    {
+        name = "progress_firmware";
```
Comment (stormc) [inline meta-iot2000-example/recipes-support/swupdate/swupdate/swupdate.cfg:11]: Well, there should be a word in the readme about this mechanism automatically rebooting after a firmware update.

File: meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend [hunk 9]
```diff
@@ -1,5 +1,23 @@
...
+do_configure_prepend () {
+        cp ${WORKDIR}/swupdate_handlers.lua ${S}
+        cp ${WORKDIR}/Makefile ${S}/tools/Makefile
+        cp ${WORKDIR}/progress_firmware.c ${S}/tools/progress_firmware.c
+}
+
+do_install_append() {
+        install -d ${D}${sysconfdir}
+        install -m 644 ${WORKDIR}/swupdate.cfg ${D}${sysconfdir}/swupdate.cfg
+        install -m 755 ${S}/tools/progress_firmware_unstripped ${D}/usr/bin/progress_firmware
```
Comment (stormc) [inline meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend]: Why the unstripped variant?

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 2 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: "sudo LANG=de_DE.UTF-8 sfdisk -d ...." also prints "last-lba"

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 3 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: The step to be done is logged in the boot log and the detailed output of the programs is logged with logger. I thought this is a good idea to separate them for cleaner kernel logs.

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 4 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: Oops

File: meta-iot2000-example/recipes-core/images/iot2000-example-image.inc (hunk 5 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-core/images/iot2000-example-image.inc]: Sorry, what do you mean by that?

File: meta-iot2000-example/README.SWUPDATE.md (hunk 6 above)
Comment (andi8086) [reply meta-iot2000-example/README.SWUPDATE.md]: true

File: meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend (hunk 9 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend]: This is because otherwise bitbake complains with QA issue that it was already stripped and will prevent future debugging. So it deploys the unstripped version in the recipe to strip it afterwards as I understand it.

File: meta-iot2000-example/recipes-support/swupdate/swupdate/swupdate.cfg (hunk 8 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-support/swupdate/swupdate/swupdate.cfg:11]: Added it in the last NOTE section.

File: meta-iot2000-example/README.SWUPDATE.md (hunk 7 above)
Comment (andi8086) [reply meta-iot2000-example/README.SWUPDATE.md]: Added

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 1 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:3]: If something fails, for example sfdisk, the script should still disable itself that's why I removed set -e. -u does not help either... if some prog fails and we i.e. set ROOT_DEVICE=, then ROOT_DEVICE is no unbound variable anymore so that -u will not cause an exit of the shell. As the script is tested already, there are no misspellings either. So I decided to remove unneeded stuff.

Comment (stormc) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:3]: Hm, if e.g. sfdisk fails, then there should be some notification to the user or something. As is, it fails silently, doesn't it?

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 3 above)
Comment (stormc) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: Hm, seems inconsistent to me to have to look in 2 places, particularly as this script is intended to run once. I'd prefer to have this in one central place.

File: meta-iot2000-example/recipes-core/images/iot2000-example-image.inc (hunk 5 above)
Comment (stormc) [reply meta-iot2000-example/recipes-core/images/iot2000-example-image.inc]: You state it's needed. OK, but *why*, i.e., what's the reason it's needed? Jut to be verbose here...

File: meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend (hunk 9 above)
Comment (stormc) [reply meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend]: If bitbake strips it, you cannot sensibly debug anymore. So, it's a matter of correctly feeding bitbake's assumptions. I would prefer to selectively skip QA here or make a note why this is here for clarity. As is, one could also think of deploying unstripped binaries to a production target image...

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 1 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:3]: I would say in our case, if sfdisk fails at that step, we have a problem with the whole system and it wont boot...

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 3 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh]: okay

File: meta-iot2000-example/recipes-core/images/iot2000-example-image.inc (hunk 5 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-core/images/iot2000-example-image.inc]: Ah sorry... I stumbled across the 'in code'. Since this is just the recipe :)

File: meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend (hunk 9 above)
Comment (andi8086) [reply meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend]: Okay I add an insane_skip

File: meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh (hunk 1 above)
Comment (stormc) [reply meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:3]: Yes, granted. Nonetheless, there should be a note left to the user that and what failed.

**Review by jan-kiszka — COMMENTED** [review]
Do we really need to duplicate all files and patches that upstream already has for 2.30?

**Comment by andi8086** [comment]
> Do we really need to duplicate all files and patches that upstream already has for 2.30?

Yes we do. Of course patches don't apply anymore and the recipe itself is very different, as is the inc file. I have tried to use a bbappend before... but we need the newer patches, the newer inc file and some changes in the bb file. And also a bbappend wont have a correct version string in its name. It is a wonder that it builds at all... without warning or error.

**Comment by jan-kiszka** [comment]
OK, different files need replacements, for sure. But there would generally be no need for bbappend hacks, see also d49d5dab3049.

Just checked: the majority of the files are identical, so please remove them. Moreover, the receipe should go into the example layer as that is the one which now needs the newer sfdisk.

**Comment by andi8086** [comment]
bbappends are no hacks but a general way to change things in an underlying recipe without duplicating files. If you say no duplication is wanted it is logical to speak of bbappends instead. Since bbappends are not possible and a lot has to be changed it is better to take the whole bunch because this is tested by upstream poky. I have the sha in the commit message where it comes from and that is poky maintained.

**Comment by jan-kiszka** [comment]
Merged into master by now
